### PR TITLE
Add PerfContext for Kv Operations

### PR DESCRIPTION
**Pull request by breeswish**
## What have you changed? (mandatory)

This PR adds PerfContext and slow log for Kv read operations. (release-2.1)

Additionally, this PR allows adjusting the threshold of outputting slow log.

## What are the type of the changes? (mandatory)

- Improvement (change which is an improvement to an existing feature)

## How has this PR been tested? (mandatory)

WIP

### CODE REVIEW

File: etc/config-template.toml [hunk 1]
```diff
@@ -111,6 +115,10 @@
 # the "scheduler too busy" error is displayed.
 # scheduler-pending-write-threshold = "100MB"
 
+# How long will a KV read request be treated as slow and print slow log.
```
Comment (zhangjinpeng1987) [inline etc/config-template.toml]: as a slow query?

File: src/coprocessor/tracker.rs [hunk 2]
```diff
@@ -136,18 +137,16 @@ impl Tracker {
...
-        }
-        if self.req_ctx.context.get_scan_detail() || is_slow_query {
-            let detail = self.total_exec_metrics.cf_stats.scan_detail();
-            exec_details.set_scan_detail(detail);
-        }
+
+        let mut handle = kvrpcpb::HandleTime::new();
+        handle.set_process_ms((time::duration_to_sec(self.item_process_time) * 1000.0) as i64);
+        handle.set_wait_ms((time::duration_to_sec(self.wait_time) * 1000.0) as i64);
+        exec_details.set_handle_time(handle);
```
Comment (zhangjinpeng1987) [inline src/coprocessor/tracker.rs]: This behavior is more frequently than before, does it impact the performance?

Comment (breeswish) [reply src/coprocessor/tracker.rs]: We always record these durations actually, so the cost in this change is only setting the structure, which should be very small.

**Comment by Connor1996** [comment]
why not add to master branch?

**Comment by breeswish** [comment]
@Connor1996 Because Meituan is using release-2.1

**Comment by breeswish** [comment]
Looks like this PR is ready for review.

File: src/storage/perf_tracker.rs [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use storage::engine::{PerfStatisticsDelta, PerfStatisticsInstant};
```
Comment (zhangjinpeng1987) [inline src/storage/perf_tracker.rs]: @breeswish I have a question for PerfStatisticsInstant, if there are multiple parallel get/scan requests in RocksDB, PerfStatisticsInstant only statistic for one operation or all the running requests?

Comment (breeswish) [reply src/storage/perf_tracker.rs]: @zhangjinpeng1987 It's thread local

**Comment by breeswish** [comment]
@zhangjinpeng1987 @Connor1996 PTAL, thanks!

File: src/coprocessor/tracker.rs [hunk 4]
```diff
@@ -74,8 +73,10 @@ impl Tracker {
     /// Initialize the tracker. Normally it is called outside future pool's factory context,
     /// because the future pool might be full and we need to wait it. This kind of wait time
     /// has to be recorded.
-    pub fn new(req_ctx: ReqContext) -> Tracker {
+    pub fn new(req_ctx: ReqContext, slow_log_threshold: Duration) -> Tracker {
         Tracker {
+            slow_log_threshold,
+
```
Comment (Connor1996) [inline src/coprocessor/tracker.rs]: extra blank line

File: src/lib.rs [hunk 5]
```diff
@@ -26,6 +26,7 @@
 #![feature(ascii_ctype)]
 #![feature(const_int_ops)]
 #![feature(use_extern_macros)]
+#![feature(optin_builtin_traits)]
```
Comment (Connor1996) [inline src/lib.rs]: Do we necessarily need this?

**Pull request closed without merging**